vio/viosslfactories.c

```diff
@@ -66,15 +66,22 @@ static unsigned char dh2048_g[]={
 static DH *get_dh2048(void)
 {
   DH *dh;
-  if ((dh=DH_new()))
-  {
-    dh->p=BN_bin2bn(dh2048_p,sizeof(dh2048_p),NULL);
-    dh->g=BN_bin2bn(dh2048_g,sizeof(dh2048_g),NULL);
-    if (! dh->p || ! dh->g)
-    {
-      DH_free(dh);
-      dh=0;
-    }
+  BIGNUM* p = BN_bin2bn(dh2048_p,sizeof(dh2048_p),NULL);
+  if (!p)
+    return 0;
+  BIGNUM* g = BN_bin2bn(dh2048_g,sizeof(dh2048_g),NULL);
+  if (!g) {
+    BN_free(p);
+  }
+  dh = DH_new();
+  if (!dh) {
+    BN_free(p);
+    BN_free(g);
+  }
+  if (!DH_set0_pqg(dh, p, NULL, g)) {
+    DH_free(dh);
+    BN_free(p);
+    BN_free(g);
```

Comment:
This part introduces double free, both `p` and `g` shouldn't be freed directly after call to `DH_set0_pqg`.
Let's maintain upstream code layout here.

Reply:
At moment, I think this code is completely dead because `DH_set0_pgq` can't fail when `p != NULL` and `g != NULL`, which we've assured at this point. I made this change because if `p == NULL` or `g == NULL`, `DH_set0_pgq` will simply return without freeing anything, so this code would be required. I extrapolated to other errors cases. Looking at the documentation, it does say it takes ownership, though.